### PR TITLE
feat: add lua patterns

### PR DESCRIPTION
**Pull request by amaanq**
Alternative to #3578 

I sort of dislike having the classes be `@keyword`, and the grammar isn't all that great...(I can fork it if anyone thinks the queries would be better handled that way since the current one uses inline/hidden rules annoyingly)

### CODE REVIEW

**Comment by clason** [comment]
maybe mark it as experimental for now, then? 

(*ceterum censeo* we _really_ need to start categorizing parsers and taking that seriously)

**Comment by amaanq** [comment]
> maybe mark it as experimental for now, then?
> 
> (_ceterum censeo_ we _really_ need to start categorizing parsers and taking that seriously)

Categorize them in what way? I do try to help out with parsers that aren't the best as much as possible..but I'm only 1 person :laughing:

**Comment by clason** [comment]
> Categorize them in what way? 

In tiers, like "core", "community maintained", "experimental", "broken" -- and then handle these differently in our CI/documentation (e.g., don't auto-update experimental or broken languages, and don't include them in the build tests etc.)

Basically, keep the list of maintainers updated, and drop maintainers (and parser tiers) aggressively.

This would also help with expectation management, so people could only install "core" languages and be confident that they work well and don't break too often.

> I do try to help out with parsers that aren't the best as much as possible..but I'm only 1 person 😆

Yes, precisely, and we need to make sure your efforts are not wasted ;)

**Comment by amaanq** [comment]
> In tiers, like "core", "community maintained", "experimental", "broken" -- and then handle these differently in our CI/documentation (e.g., don't auto-update experimental or broken languages, and don't include them in the build tests etc.)

Yeah that is a **really** good idea, and could help with the CI breaking ever so often..would "core" be those at tree-sitter? And say a self-made parser (by me or anyone else but NOT at tree-sitter) be "community maintained"?

**Comment by amaanq** [comment]
Here's some screenshots anyways, one thing I kept in mind was keeping consistency with Regex highlights:

image

image

image

**Comment by clason** [comment]
> Yeah that is a **really** good idea, and could help with the CI breaking ever so often..would "core" be those at tree-sitter? And say a self-made parser (by me or anyone else but NOT at tree-sitter) be "community maintained"?

No, those would be parsers that nvim-treesitter core maintainers (like you ;)) care about enough to keep running (or that are stable enough that no maintenance is necessary).

"community" would be those maintainers we always need to ping manually (which could be automated as well).

(TL;DR: it's about maintaining the _queries_ here, not the parsers upstream.)

**Comment by amaanq** [comment]
Man that windows ci is taking forever...typical

Feel free to merge whenever so that your neovim PR can be updated accordingly :)  (as long as class being captured as `@keyword` is ok to you)

**Comment by amaanq** [comment]
Text inside brackets weren't highlighted as `@constant` (e.g. A and Z in A-Z), classes inside brackets weren't highlight as `@keyword` because for whatever reason the inner class node is hidden (_class) and is used there, and I added fixes to only inject inside the string content and inject in both parameters for gsub/gmatch (both can have luap but before it was only injecting into the first), and maybe some other things I forgot

**Review by ObserverOfTime — APPROVED** [review]

**Comment by amaanq** [comment]
Switched to my fork, queries look nicer and capture better as well

**Comment by amaanq** [comment]
cc @max397574 iirc you expressed interest in this

File: queries/luap/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  "-" @punctuation.delimiter
+  to: (character) @constant)
+
+(set
+  (character) @constant)
+
+(class) @keyword
+
+(negated_set
+  "^" @string.regex
```
Comment (ObserverOfTime) [inline queries/luap/highlights.scm:31]: I'd make this one `@operator` too.
```suggestion
  "^" @operator
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

**Review by ObserverOfTime — APPROVED** [review]

**Pull request merged**